### PR TITLE
Observability & Dashboard Stack

### PR DESCRIPTION
**Pull request by vinayakg**
## 🎯 **Overview**

This PR introduces a **complete self-hosted observability and dashboard stack** for monitoring and managing large-scale LLM training operations (70B+). It replaces SaaS tools with high-performance, local-first components and provides a **production-ready React dashboard** for real-time and historical data visualization.

**Key Scope:**
- ✅ **Backend Observability Stack** (ClickHouse + Vector + TrainingOps)
- ✅ **Frontend Dashboard Stack** (React + TypeScript + Vite)
- ✅ **Integration Layer** (Aggregation API + Metrics Server)
- ✅ **Security** (HTTPS, TLS, Role-based access, IAM cross-account, AWS Secrets Manager)

---

## 📊 **Architecture Overview (Full Stack)**
```
┌─────────────────────────────────────────────────────────────────────────────┐
│                          TRAINING INSTANCE (Compute)                        │
├─────────────────────────────────────────────────────────────────────────────┤
│                                                                             │
│  TrainingOps (Facade)                                                       │
│    ├─ JSONLogger ──────┐                                                    │
│    ├─ SystemMetrics ───┤ JSONL Logs                                         │
│    ├─ MetricsServer    │  (Port 8000)                                       │
│    └─ CheckpointRegistry                                                    │
│                        │                                                    │
│  ┌─ Vector Sidecar (systemd service) ────────────────────────────────────┐  │
│  │   EC2 Userdata Provisioning:                                         │  │
│  │   • Installed & auto-started as systemd unit on instance boot        │  │
│  │   • ClickHouse password + endpoint fetched from AWS Secrets Manager  │  │
│  │     (no .env files copied, no secrets shared with users)            │  │
│  │   • Transforms (JSONL → typed table routes):                         │  │
│  │     - to_raw_logs                                                   │  │
│  │     - to_metric_points                                              │  │
│  │     - to_checkpoints                                                │  │
│  │                                                                      │  │
│  │ /tmp/training_logs/*.jsonl  ─── (read by Vector)                    │  │
│  └─────────────────────────��────────────────────────────────────────────┘  │
│                                                                             │
│                        ──────► HTTPS                                        │
└─────────────────────────────────────────────────────────────────────────────┘
                              │ :8443
                              │ (TLS + Auth)
                              │
┌───────────────────────────────▼────────────────────────────────────────────┐
│                     DB INSTANCE (Data Warehouse)                           │
├────────────────────────────────────────────────────────────────────────────┤
│                          ClickHouse                                        │
│                      (HTTPS :8443)                                         │
│                                                                            │
│  Credential Management:                                                   │
│  • DB password & endpoint retrieved from AWS Secrets Manager              │
│  • Cross-account IAM role assumption for secrets access                   │
│  • No hardcoded credentials, no .env distribution                         │
│                                                                            │
│  Tables:                           Users:                                 │
│  • logs                            • p12_writer (Vector, INSERT+SELECT)   │
│  • metric_points                   • p12_reader (Dashboard, SELECT only)  │
│  • metric_arrays                   • default (ClickHouse admin, localhost)│
│  • checkpoints                                                             │
│  • events                                                                  │
│  • runs                                                                    │
└────────────────────────────────────────────────────────────────────────────┘
                                  │
              ┌───────────────────┴──────────────────┐
              │                                      │
┌───────────────▼──────────────────┐  ┌──────────────▼─────────────────────┐
│  DASHBOARD BACKEND (Aggregator)  │  │  DASHBOARD FRONTEND (React/Vite)  │
├──────────────────────────────────┤  ├───────────────────────────────────┤
│ FastAPI (Port 8081)              │  │  React + TypeScript + Vite        │
│                                  │  │  • Component-based UI             │
│ GET /runs                        │  │  • Real-time metric rendering     │
│ GET /metrics                     │  │  • Historical data visualization  │
│ GET /metric/history              │  │  • Checkpoint management UI       │
│                                  │  │  • Run status dashboard           │
│ Merges:                          │  │  • System metrics display         │
│ • Hot (MetricsServer < 1h)       │  │  • Training progress tracking     │
│ • Cold (ClickHouse > 1h)         │  │  • Alert/event viewer            │
│                                  │  │                                   │
│ Pydantic models for type safety  │  │  ESLint + TypeScript strict mode │
│ Mock adapters for demo mode      │  │  Tailwind CSS styling             │
└──────────────────────────────────┘  └───────────────────────────────────┘
```
## 🔧 **Backend Components** (7 Core Pieces)

| # | Component | File | Role |
|---|-----------|------|------|
| 1 | **TrainingOps Facade** | `training_ops.py` | Single entry point for training loop integration |
| 2 | **JSONLogger** | `train_logger/json_logger.py` | Non-blocking structured logging → JSONL files |
| 3 | **SystemMetricsCollector** | `system_metrics/collector.py` | System metrics (CPU/RAM/GPU/Net) → JSONL |
| 4 | **MetricsServer** | `metrics_server.py` | Custom HTTP API (Port 8000) for live metric queries |
| 5 | **CheckpointRegistry** | `checkpoint_registry/checkpoint_registry.py` | ClickHouse-backed checkpoint governance |
| 6 | **Vector Sidecar** | `sidecar_agent/vector.toml` | Data shipper (JSONL → ClickHouse via HTTPS). Provisioned via systemd + EC2 userdata. Credentials from Secrets Manager. |
| 7 | **Watchdog Service** | `watchdog/watchdog.py` | Control plane for anomaly detection & training pause |

---

### **Technology Stack**
- **Framework:** React 19 with TypeScript (strict mode)
- **Build Tool:** Vite with HMR (Hot Module Replacement)
- **Styling:** Tailwind CSS (pre-configured)
- **Linting:** ESLint with React plugins
- **Package Manager:** npm/yarn

#### **1. Core UI Components**
- **Run List View** — Browse all active/completed training runs with metadata
- **Metrics Dashboard** — Real-time loss, learning rate, throughput, gradient norms
- **System Metrics Display** — CPU, memory, GPU utilization, disk, network stats
- **Checkpoint Browser** — View, filter, and manage checkpoints with protection status
- **Event Log Viewer** — Training events (OOM, divergence, stage transitions)
- **Training Progress** — Step-based progress bars, ETA, tokens/sec tracking

#### **2. Data Visualization**
- **Time-series charts** for metrics (loss curves, LR schedule)
- **System resource graphs** (CPU/GPU utilization over time)
- **Checkpoint timeline** with loss annotations
- **Live metric updates** (sub-second latency via WebSocket or polling)
- **Historical drill-down** (zoom to specific time ranges)

#### **3. UI Integration Points**
```typescript
// Example component architecture
src/
├── components/
│   ├── RunList.tsx           // Fetches from GET /runs
│   ├── MetricsChart.tsx      // Calls GET /metrics?run_id=X&metric=loss
│   ├── CheckpointBrowser.tsx // Renders checkpoint table
│   ├── SystemMetrics.tsx     // Displays CPU/GPU gauges
│   ├── EventLog.tsx          // Shows training events
│   └── TrainingStatus.tsx    // Overall progress & ETA
│
├── api/
│   └── dashboardClient.ts    // HTTP client for Aggregation API
│
├── hooks/
│   ├── useRunQuery.ts        // Fetch & cache run data
│   ├── useMetricsHistory.ts  // Fetch metric time-series
│   └── useAutoRefresh.ts     // Polling/WebSocket for live updates
│
└── pages/
  ├── Dashboard.tsx         // Main landing page
  ├── RunDetail.tsx         // Single run deep-dive
  └── Settings.tsx          // Dashboard config (refresh rates, etc.)
```

#### **4. Key Files Added**
| File | Purpose |
|------|---------|
| `monitoring-dashboard/src/main.tsx` | React app entry point |
| `monitoring-dashboard/src/App.tsx` | Root component with routing |
| `monitoring-dashboard/src/components/**` | Reusable UI components |
| `monitoring-dashboard/src/api/dashboardClient.ts` | Aggregation API client |
| `monitoring-dashboard/vite.config.ts` | Vite build configuration |
| `monitoring-dashboard/tsconfig.json` | TypeScript strict mode config |
| `monitoring-dashboard/tailwind.config.js` | CSS utility framework |
| `monitoring-dashboard/eslint.config.js` | Linting rules (React + TS) |
| `monitoring-dashboard/package.json` | Dependencies (React, Vite, etc.) |

### **Backend API** (`aggregation_api/dashboard_backend.py`)
- **Framework:** FastAPI with Pydantic models
- **Port:** 8081
- **Endpoints:**
- `GET /runs` — List all training runs
- `GET /metrics?run_id=X&metric=loss&window=1h` — Fetch metric data (hot + cold)
- `GET /metric/history?run_id=X&metric=loss` — Historical deep-dive

### **Data Fusion Strategy**
```
User requests metric data:
│
├─► Query MetricsServer (Port 8000) for LAST 1 HOUR (hot, fast, accurate)
│
└─► Query ClickHouse for DATA > 1 HOUR OLD (cold, compressed, historical)

Result: Seamless single time-series to frontend (no client-side merging needed)
```

---

### **Vector Service Provisioning**
- ✅ **Systemd Service** — Installed and auto-launched as a managed systemd unit on instance boot.
- ✅ **EC2 Userdata Bootstrap** — Training instance startup script (`components/sidecar_agent/userdata_vector.sh`) configures Vector:
- Installs Vector binary & dependencies
- Registers as systemd service (enables auto-restart, proper logging, lifecycle management)
- Fetches ClickHouse DB password & endpoint from **AWS Secrets Manager** at boot
- No manual setup, no SSH required, fully hands-off for users
- ✅ **Secrets Manager Integration** — ClickHouse credentials are retrieved programmatically via AWS SDK (STS assume-role for cross-account access)

### **Observability Stack Security**
- ✅ **HTTPS only** (Port 8443) with self-signed or custom TLS certs
- ✅ **HTTP localhost-only** (Port 8123) for admin/healthchecks
- ✅ **Role-based access:**
- `p12_writer` — Vector, training instance (INSERT+SELECT)
- `p12_reader` — Dashboard, dashboard teams (SELECT only)
- `default` — ClickHouse admin (localhost only)
- ✅ **No Secret Sprawl** — Credentials never hardcoded, never in `.env` files, never distributed to users
- ✅ **Cross-account IAM** for Vector bootstrap (training account → infra account)
- ✅ **TLS certs, password hashes, `.env` files excluded from git** (`.gitignore`)

### **Dashboard**
- ✅ **Read-only user access** (p12_reader credentials)
- ✅ **Protected checkpoints** (tags: `growth`, `lora`, `release_candidate`)
- ✅ **Audit trail** via ClickHouse `logs` table

---

### **.pre-commit-config.yaml**
```yaml
# Added exclusion for TypeScript config JSON files
- id: check-json
exclude: tsconfig.*\.json$
```
(Prevents pre-commit from validating JSON comments in `.tsx` files)

## 🧪 **Testing Coverage**

| Component | Test File | Status |
|-----------|-----------|--------|
| JSONLogger | `components/tests/test_logger.py` | ✅ Context merging verified |
| MetricsServer | `components/tests/test_metrics_server.py` | ✅ HTTP endpoints tested |
| CheckpointRegistry | (Coming soon) | ⏳ ClickHouse queries |
| Dashboard UI | (React testing library) | ⏳ Component rendering |

---

### **For Backend Reviewers:**
1. **ClickHouse Schema** — See `components/clickhouse/initdb.d/`
  - `logs` table: audit trail with metrics/context as JSON
  - `metric_points` table: typed scalars (one row per metric)
  - `checkpoints` table: ReplacingMergeTree for governance
2. **Vector Config** — See `components/sidecar_agent/vector.toml`
  - Verify transforms (JSON parsing → typed table routes)
  - **EC2 Bootstrap Script** — `components/sidecar_agent/userdata_vector.sh`
    - Section 1-2: Vector installation
    - Section 3-4: Systemd service registration
    - Section 5-6: Secrets Manager credential retrieval
3. **TrainingOps Integration** — See `components/training_ops.py`
  - Simple API surface for training teams

### **For Frontend Reviewers:**
1. **React Structure** — `monitoring-dashboard/src/`
  - Component hierarchy and prop drilling
  - TypeScript strict mode compliance
2. **API Client** — `aggregation_api/dashboard_backend.py` + `src/api/dashboardClient.ts`
  - Request/response types (Pydantic ↔ TypeScript)
3. **Styling** — Tailwind CSS utility coverage
  - Responsive design (mobile → desktop)
4. **Linting** — ESLint + TypeScript rules in strict mode

### **For Infrastructure Reviewers:**
1. **Vector Bootstrap Automation** — See `components/sidecar_agent/userdata_vector.sh`
  - Complete EC2 userdata flow (install → configure → launch)
  - Secrets Manager integration (no hardcoded credentials)
2. **Cross-Account IAM** — See `components/clickhouse/Post_Clickhouse_Install.md`
  - Section 4: `setup-training-account.sh` + `setup-infra-account.sh`
  - STS assume-role policies for Secrets Manager access
3. **EBS & Snapshots** — See `components/clickhouse/Post_Clickhouse_Install.md`
  - Section 2 & 5: Volume attachment, DLM policy

## 📚 **Resources**

- **[ClickHouse Setup Guide](experiments/12_training_operations/components/clickhouse/SETUP.md)** — DB initialization & security
- **[Post-Install Checklist](experiments/12_training_operations/components/clickhouse/Post_Clickhouse_Install.md)** — AWS security groups, Secrets Manager, EBS, cross-account IAM
- **[Vector Bootstrap Script](experiments/12_training_operations/components/sidecar_agent/userdata_vector.sh)** — EC2 userdata automation for systemd setup & credential retrieval
- **[Component Integration Guide](experiments/12_training_operations/components/README.md)** — Training team quick-start
- **[System Architecture](experiments/12_training_operations/components/system_architecture.md)** — Data flow & control plane
- **[Troubleshooting](experiments/12_training_operations/monitoring-dashboard/TROUBLESHOOT.md)** — Common issues & debugging

## Reviewers

- [ ] **Reviewer 1**: A member from your own team.
- [ ] **Reviewer 2**: A member from the repo owners team (@The-School-of-AI/llm-repo-owners). 

> **Note**: Every pull request requires atleast 2 reviewers/approvers before it can be merged.

### CODE REVIEW

**Review by abhi1021 — COMMENTED** [review]
Based on the review I just completed, here are the **high severity errors** from PR #527:

## 🔴 High Severity Issues

### 1. `window` Query Parameter Silently Ignored — `dashboard_backend.py`
The `/metrics` endpoint accepts a `window` parameter (e.g., `?window=24h`) but `cutoff_time` is always hardcoded to `now - 300` (5 minutes) . Every caller requesting a different time window will silently receive incorrect 5-minute data with no warning or error — a correctness bug that's hard to debug in production.

**Fix:** Parse `window` into seconds and apply it:
```python
window_map = {"1h": 3600, "24h": 86400, "7d": 604800}
cutoff_time = now - window_map.get(window, 3600)
```

***

### 2. No Error Handling for Invalid Inputs — `dashboard_backend.py`
There are no guards for invalid `run_id` or unknown `metric` names . The API silently returns mock/empty data for any input rather than raising an `HTTPException` (which is ironically already imported but unused). In a live monitoring dashboard for a 70B training run, this means a misconfigured client gets no feedback that it's querying the wrong metric.

**Fix:** Add explicit 404 checks:
```python
if run_id not in valid_runs:
    raise HTTPException(status_code=404, detail=f"run_id '{run_id}' not found")
```

***

### 3. Checkpoint Exception Swallowed by Late Import — `checkpoint_manager.py`
`get_metrics_server` is imported *inside* the `except` block . If this import fails (e.g., circular import, missing module), the **original checkpoint exception is completely lost** and replaced with an unrelated `ImportError`. During a multi-hour 70B training run, this could make a real checkpoint failure appear as a metrics infrastructure error, sending the team chasing the wrong root cause.

**Fix:** Move the import to the top of the file and wrap the metrics call in its own `try/except` so the original exception is always re-raised via `finally`:
```python
try:
    record_checkpoint(step, success=False)
except Exception:
    pass  # don't let metrics failure mask checkpoint failure
finally:
    raise  # always re-raise original
```

**Review by abhi1021 — APPROVED** [review]
✅ Issue #3 — Exception masking (checkpoint_manager.py) — FULLY FIXED

The diff clearly shows get_metrics_server moved to line 8 at module level (green +), and both late imports inside the try and except blocks removed (red - on lines 70 and 78). This is exactly the correct fix.
​

✅ Issues #1 & #2 — dashboard_backend.py deleted

**Review by aiplaybookin — APPROVED** [review]

**Review by pankajkr23 — APPROVED** [review]
Approved. Thanks!
